### PR TITLE
HPCC-19862 include Spark package

### PR DESCRIPTION
**Pull request by xwang2713**
Spark package is deployed under <HPCC Install directory>/external/
Spark-HPCC jar file is deployed under <HPCC Install directory>/jar/spark/
CMake options:
-DUSE_SPARK=ON : this will trigger build Platform or LN to include Spark

-DSPARK_URL=<value>: This is a URL to download spark/hadoop package. The default is "http://apache.mirrors.pair.com/spark/spark-2.3.1/spark-2.3.1-bin-hadoop2.7.tgz". In local build we use http://10.240.32.242/data3/software/spark/spark-2.3.1-bin-hadoop2.7.tgz

-DSHA512=<value>. It is for verify downloaded package. The default value is for spark-2.3.1-bin-hadoop2.7.gz. User can disable by set '-DSHA512=OFF" or "-DSHA512=", though it is no recommended.

-DSPARK_URL_BASE: The defaulit is "http://apache.mirrors.pair.com"

-DSPARK_VERSION: The default is "2.3.1"
-DHADOOP_VERSION: The default is "2.7"
When "SPARK_URL" is provided the versions will be parsed from the URL.

-DSPARK_HPCC=<value> : spark-hpcc jar file. The path will be related to build directory.

Signed-off-by: Xiaoming Wang <xiaoming.wang@lexisnexi.com>

@Michael-Gardner  @rpastrana please review

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  

## Testing:

http://10.240.32.243/view/Spark/job/Spark-Package/

### CODE REVIEW

File: spark/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,110 @@
+################################################################################
+#    HPCC SYSTEMS software Copyright (C) 2016 HPCC Systems.
```
Comment (rpastrana) [inline spark/CMakeLists.txt]: 2018

File: spark/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+#########################################################
+# Description:
+# ------------
+#    Spark with HPCC
+#
+#
+#########################################################
+cmake_minimum_required(VERSION 3.3)
+
+project(spark)
```
Comment (rpastrana) [inline spark/CMakeLists.txt]: let's call this project and its folder spark-integration

File: spark/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+     CODE "
+     execute_process( WORKING_DIRECTORY \"${CMAKE_BINARY_DIR}/downloads\" COMMAND \"${CMAKE_COMMAND}\" -E tar -zxf \"${SPARK_PACKAGE_NAME}.tgz\" )
+     execute_process( WORKING_DIRECTORY \"${CMAKE_BINARY_DIR}/downloads\" COMMAND \"${CMAKE_COMMAND}\" -E remove -f  \"${SPARK_PACKAGE_NAME}.tgz\" )
+    "
+  )
+
+  install(
+    DIRECTORY "${CMAKE_BINARY_DIR}/downloads/${SPARK_PACKAGE_NAME}"
+    COMPONENT runtime
+    DESTINATION "external"
```
Comment (rpastrana) [inline spark/CMakeLists.txt]: externals

File: spark/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    install(CODE "file(
+      DOWNLOAD ${SPARK_URL}
+      ${CMAKE_BINARY_DIR}/downloads/${SPARK_PACKAGE_NAME}.tgz
+      TIMEOUT 120
+      INACTIVITY_TIMEOUT 30)"
+    )
+  endif (SHA512STRING)
+
+  install(
+     CODE "
```
Comment (rpastrana) [inline spark/CMakeLists.txt]: do we need explicit uninstall logic?

**Review by rpastrana — COMMENTED** [review]
@xwang2713 a couple of comments

File: spark/CMakeLists.txt (hunk 4 above)
Comment (xwang2713) [reply spark/CMakeLists.txt]: not really, deb package should remove everything we layout during installation

File: spark/CMakeLists.txt (hunk 1 above)
Comment (xwang2713) [reply spark/CMakeLists.txt]: Will do

File: spark/CMakeLists.txt (hunk 2 above)
Comment (xwang2713) [reply spark/CMakeLists.txt]: OK

**Comment by xwang2713** [comment]
rebase and updated.

File: spark/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,110 @@
+################################################################################
```
Comment (rpastrana) [inline spark/CMakeLists.txt]: does it make sense to name the folder spark-integration?

**Review by rpastrana — COMMENTED** [review]
@xwang2713 just one minor comment

File: spark/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  )
+
+  install(
+    DIRECTORY "${CMAKE_BINARY_DIR}/downloads/${SPARK_PACKAGE_NAME}"
+    COMPONENT runtime
+    DESTINATION "externals"
+  )
+
+  install(
+    FILES "${CMAKE_BINARY_DIR}/${SPARK_HPCC}"
```
Comment (rpastrana) [inline spark/CMakeLists.txt]: @Michael-Gardner this implies we can provide a mvn build target of spark-hpcc(no deps + wsclient-withdeps)

**Comment by xwang2713** [comment]
One more thing need to consider: currently i name the spark directory as externals/<spart package name with version>.  Ideally it should have common name with symbolic link. Not sure if all Linux packages (rpm, deb,etc) can handle symbolic link correctly. The other choice is to name the directory as common name directly, for example spark-hadoop. Inside "RELEASE" file will tell the version information.

File: spark/CMakeLists.txt (hunk 5 above)
Comment (rpastrana) [reply spark/CMakeLists.txt]: @xwang2713 does it make sense to rename the folder 'spark' to 'spark-integration'?

File: spark/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  message("")
+  message("SPARK VERSION: ${SPARK_VERSION}")
+  message("HADOOP VERSION: ${HADOOP_VERSION}")
+  message("SPARK Package URI: ${SPARK_URL}")
+  message("")
+
+  if(SHA512)
+     set( SHA512STRING "${SHA512}" )
+  else(SHA512)
+     set( SHA512STRING "DC3A97F3D99791D363E4F70A622B84D6E313BD852F6FDBC777D31EAB44CBC112CEEAA20F7BF835492FB654F48AE57E9969F93D3B0E6EC92076D1C5E1B40B4696" )
```
Comment (rpastrana) [inline spark/CMakeLists.txt]: does it make sense to pull this dynamically or would that defeat the purpose?

File: spark/CMakeLists.txt (hunk 6 above)
Comment (rpastrana) [reply spark/CMakeLists.txt]: @Michael-Gardner can you comment on the contents of spark-hpcc's jar?

**Comment by rpastrana** [comment]
@xwang2713 per our conversation, it makes sense to extract to /opt/HPCCSystems/externals/spark-hadoop. This gives @Michael-Gardner a stable target for the dynamic configuration. The version of the spark instance can be determined by the content of the /opt/HPCCSystems/externals/spark-hadoop/RELEASE file.

**Comment by xwang2713** [comment]
Change unpacked spark name to "spark-hadoop".

**Comment by xwang2713** [comment]
This pull request will be covered in HPCC-20169

**Pull request closed without merging**